### PR TITLE
client-go: support waiting for SharedInformerFactory shutdown

### LINKED ISSUE #65036: SharedInformer documentation is missing or hard to find
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
I cannot find documentation on the correct way to use informers & shared informers. Without guidance, I ended up writing code that called `"k8s.io/client-go/informers/core/v1".NewPersistentVolumeInformer` (and some others). Although I read the docstring of that method, it was only through conversation with a more experienced contributor that I learned I should instead be calling `"k8s.io/client-go/informers".(*sharedInformerFactory).Core().V1().PersistentVolumes().Informer()`

I also had difficulty understanding whether I needed to call `Run()` on each informer, or on the informer factory. The correct answer turns out to be
`"k8s.io/client-go/informers".(*sharedInformerFactory).Start()`
but that was not apparent from the godoc as `Start()` comes from an embedded interface which I didn't notice.

EDIT: Calling `Start()` before creating informers in the factory results in those informers never generating events. I suppose this means that one should never call `Start()` until all informers have been created? Not sure.

**What you expected to happen**:
There should be a doc somewhere that says, "To use a shared informer, make a factory and call `Start()` on it." and that doc should be easy to find.

/sig api-machinery

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

SharedInformerFactory starts goroutines in Start and those can be stopped by
closing the stop channel. However, there was no API that waits for the
goroutines.

This is a problem for unit testing. A test has to return while the informers
are still running, which may get flagged by tools like
https://github.com/uber-go/goleak or by klog/ktesting when those informers
lead to log output.

While at it, more documentation gets added to address
https://github.com/kubernetes/kubernetes/issues/65036.

#### Which issue(s) this PR fixes:

Fixes #65036 (as confirmed in https://github.com/kubernetes/kubernetes/issues/65036#issuecomment-1238731026)

#### Does this PR introduce a user-facing change?
```release-note
client-go: SharedInformerFactory supports waiting for goroutines during shutdown
```

/cc @alexzielenski

### CODE REVIEW

File: staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go
Comment (alexzielenski) [inline staging/src/k8s.io/apiextensions-apiserver/examples/client-go/pkg/client/informers/externalversions/factory.go:190]: I think example code would be helpful than descriptions for this. Could you add an example instead

**Review by alexzielenski — CHANGES_REQUESTED** [review]
looks great so far. Would like to see the semantics for `WaitForShutdown` clarified a bit more in the documentation.

I'm also wondering what the best place would be to add tests for this. I'm not seeing any existing tests for the imformer-gen code. Hopefully an approver can comment about that.

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:198]: This is the same as:
```
factory := NewSharedInformerFactory(....)
_ := factory.InformerFor(gvk1)
factory.Start(stop1)
_ := factory.InformerFor(gvk2)
factory.Start(stop2)
close(stop1)
factory.WaitForShutdown()
```

In other words, it will hang until some other goroutine closes `stop2`.

I need to change the description to "WaitForShutdown waits for the factory to have no goroutines running anymore."

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:260]: The advantage of embedding this in the generated code is that all factories will have it in their godoc description.

If I were to add an example, where would it go? Should it replace the doc comment?

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:187]: No, it's worse: without it, when the goroutines run, they uses the current value of the loop variable, typically the last entry in the slice.

This idiom creates a new variable for each loop iteration. I can add a comment explaining it.

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (alexzielenski) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:260]: Right I would add a small example of the correct usage in the doc code block comment

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (alexzielenski) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:187]: shucks thats unfortunate. maybe a brief line about it yeah

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:260]: Got it. I thought initially that you meant an `ExampleSharedInformerFactory` function in a separate _test.go file. That would have been more complicated.

As I am now touching the documentation, let me make some other enhancements:
- I am moving the comments for the functions from the implementation to the interface, because then they appear in the godoc page.
- I am adding methods that are "hidden" in the internal interface to the user-visible SharedInformerFactory, in particular `Start`. That way users see everything they need to know in one godoc page.

**Comment by pohly** [comment]
@alexzielenski : please take another look. I prefer to keep the PR always mergeable, therefore I updated the commit. Changes since initial review: https://github.com/kubernetes/kubernetes/compare/64834e96aa00e041717c110e9ff051b1f8ddf482..5ffd0887ed858e01d106074314b759695afffe3e

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (alexzielenski) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:281]: I'm not sure this doc-comment gives enough attention to the case we discussed earlier in this thread: https://github.com/kubernetes/kubernetes/pull/112200#discussion_r961758712

**Review by alexzielenski — APPROVED** [review]
nit about the doc comment

otherwise looks good to me. ill hold off on proper /lgtm until someone more acquainted takes a look.

xref: https://kubernetes.slack.com/archives/C0EG7JC6T/p1662030952291459

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:281]: There I suggested "WaitForShutdown waits for the factory to have no goroutines running anymore." Isn't "until all goroutines have stopped" saying the same, just with fewer words?

Comment (alexzielenski) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:281]: I suppose. Personally I read the two differently:

"all goroutines have stopped" - does this refer to all goroutines at start of function call or at the time of shutdown? My initial read is start of function call.

"no goroutines are running anymore" - this communicates more directly to me that at the time of stopping, there will be no goroutines running. But that is just my read of it. You are right others might not find that sufficient also.

Maybe that is good reason enough to expound on it in a few more words. To me it is an important detail to get across: if different stop channels are used for the multiple calls to `Start`, the semantics for this function change.

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:281]: > "all goroutines have stopped" - does this refer to all goroutines at start of function call or at the time of shutdown?

There is no "time of shutdown". When `WaitForShutdown` is called, it checks for running goroutines and blocks until those terminate.

> if different stop channels are used for the multiple calls to Start, the semantics for this function change.

I don't see the difference. "all goroutines" includes "all" goroutines, regardless how they were started. So does "no goroutines are running".

Let's hear from someone else. I can change the wording, I just don't see why and would like to understand that first.

Comment (alexzielenski) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:281]: > There is no "time of shutdown". When WaitForShutdown is called, it checks for running goroutines and blocks until those terminate.

Yes, and if any new goroutines are added while blocking, `WaitForShutdown` will wait for those too. This is a huge difference from just waiting for routines that existed before blocking. The current doc-comment does not communicate this nuance.

> I don't see the difference. "all goroutines" includes "all" goroutines, regardless how they were started. So does "no goroutines are running".

Sure, then perhaps something different from those wordings.

> Let's hear from someone else. I can change the wording, I just don't see why and would like to understand that first.

👍🏼

Comment (lavalamp) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:281]: The question I'd answer in the comment is whether new goroutines can start or them all having stopped is a permanent condition. "Shutdown" implies that it's permanent, but both proposed words don't demand this interpretation.

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: So the problem is not the function *comment*, but the function *name*. How about this (pushed):

	// WaitForStop blocks until all goroutines have terminated. For that
	// the close channel(s) that they were started with must be closed,
	// either before WaitForStop gets called or while it is waiting.
	// It is possible to start and stop multiple times.
	WaitForStop()

Comment (lavalamp) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Yeah, that's better! It raises some followup questions, mostly do we really recommend that people call this multiple times and/or with different stop channels?

From an implementation point of view, I think it may be impossible to use this safely (if there are multiple starts that the author of the call to WaitForStop() is unaware of) since the code is using a WaitGroup and: https://pkg.go.dev/sync#WaitGroup.Add

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: > mostly do we really recommend that people call this multiple times and/or with different stop channels?

The existing API allows that, so in that sense I am only documenting what people may have already started to rely upon. I don't think we should disallow it now.

> it may be impossible to use this safely

Hmm, what can go wrong?

With "multiple times" I meant:
```
factory.Start(...)
# close channel
factory.WaitForStop()

factory.Start(...)
# close channel
factory.WaitForStop()
```

That is safe. What is non-deterministic is calling Start and WaitForStop in different goroutines without coordination. The WaitForStop may have already returned when Start increments the counter again.

The WaitGroup docs says: "Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.".

I think this just means that `wg.Wait()` followed by `wg.Add(1)` will not work because `wg.Wait()` returns immediately. But that's consistent with how `WaitForStop` is defined: if it gets called before `Start`, then no goroutines were started and thus (logically true) "all goroutines (that were started so far) have terminated".

Comment (lavalamp) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Right. If the caller doesn't have any synchronization, they could call WaitForStop() and as of that moment everything is stopped, but by the time they do their next line (whatever that is) a new goroutine could have started again. Functions like this usually don't require synchronization like that because usually being stopped is a permanent condition. If it's not permanent, it's potentially hard to use this safely, depending on what the caller is going to do next.

Why are people going to call WaitForStop? The only reason that comes to mind is that we're going to shut down completely and don't want undrained queues or extra goroutines.

My first thought on how to make this safe is to just make it permanent:
* Change the name to just Stop() or Shutdown()
* When called, set a "shuttingDown" flag to true
* Make start/run refuse to do so if the flag is true
* Document that caller must also close all stop channels (alternatively, you could store them and close them for the caller, but that seems like maybe too much work)

This won't harm existing callers because there's no shutdown method at all. Your intended use case of test cases will never need to re-start a stopped thingy.

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: The use case is indeed mostly unit tests.

A `Shutdown` function with a behavior as described works for me, there's just one problem: should `Start` then be changed to return an error (currently it has none) or just silently not start informers?

I suppose it has to be the latter, because adding an error would force everyone to add error checking to avoid linter errors for a situation that never occurs in their code.

Comment (lavalamp) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Yeah, that's a dilemma...

One more option is to panic in that condition.

I guess yeah, make the existing `Start` silently ignore this. Add on to the comment...

I don't feel great about that because it'll be hard to notice if it does happen to someone, but the other options seem to involve more work for more people.

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: I pushed a version with `Shutdown`.

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:217]: This can be "shuttingDown", and you can set it to true at the beginning. Then you don't have to hold the lock the whole time, and "Start" likewise can return immediately instead of blocking until Shutdown releases the lock.

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:217]: I had considered that, but then the behavior of Shutdown changes slightly: when one goroutine is blocked in Shutdown waiting for goroutines, then a second call to it in another goroutine has to return immediately because there is nothing it can wait on. It cannot block on the mutex (not locked while waiting), and it cannot call `f.wg.Wait` (may only be called by one goroutine at a time).

Is that okay?

Comment (lavalamp) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:217]: > cannot call f.wg.Wait (may only be called by one goroutine at a time).

Are you sure? I don't see that in the godoc for Wait()? I have let multiple goroutines call it at the same time in various places.

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:217]: I saw this in the source code:

https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/sync/waitgroup.go;l=140-142

> panic("sync: WaitGroup is reused before previous Wait has returned")

Can that be ignored?

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:217]: I tried to trigger that panic, but without success. Perhaps I was too cautious and read more into that panic than is warranted. I'm not sure what is meant with "reused". :shrug:

Comment (lavalamp) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:217]: My guess is that's internal housekeeping, but if it's not we should still be fine, since setting the flag & checking it before adding will prevent any more calls of Add, so there's no chance of racing between Add and Wait. Right?

Comment (pohly) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go]: Agreed. I've updated the PR to use `shuttingDown` and not hold the mutex while waiting for goroutines.

**Comment by pohly** [comment]
/retest

Because of https://github.com/kubernetes/kubernetes/issues/112388

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:298]:
```suggestion
	// In addition, Shutdown blocks until all goroutines have terminated. For that to happen,
```

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:302]:
```suggestion
	// Shutdown may be called multiple times, even concurrently. All such calls will
```

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:303]:
```suggestion
	// block until all goroutines have terminated.
```

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (lavalamp) [inline staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:209]: it really doesn't matter here, but for more complex things I usually ask people to put it in an anonymous function so that `defer` can be used to unlock.

File: staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go
Comment (lavalamp) [reply staging/src/k8s.io/code-generator/cmd/informer-gen/generators/factory.go:196]: This is a common go gotcha-- I had to think for a moment to see why the old version of the code didn't have this bug.

**Comment by lavalamp** [comment]
Thanks!

/approve

your choice about which nits to fix

**Comment by pohly** [comment]
I took your documentation updates but left the Lock/Unlock as it was. I think it's simple enough to not need a separate function.

**Comment by pohly** [comment]
@lavalamp: do you want to at the LGTM yourself or hand it back to @alexzielenski?

**Review by alexzielenski — APPROVED** [review]
/lgtm

**Comment by howardjohn** [comment]
Are there any plans to add this for metadata SharedInformerFactory and dynamic DynamicSharedInformerFactory?

**Comment by pohly** [comment]
No plans, but if someone (you?) were to prepare a PR, it probably would get merged.

**Pull request merged**
